docs/src/explanations/properties.md

```diff
@@ -148,3 +148,249 @@ on integral `Property` typed values.
+coexist within the hardware instance graph. `Object` instances can exist
+within hardware modules, providing domain-specific information, but hardware
+instances cannot exist within `Class` declarations.
+
+`Object`s can be referenced, and references to `Object`s are a special kind of
+`Property[ClassType]` type. This allows the data model captured by `Class`
+declarations and `Object` instances to form arbitrary graphs.
+
+To understand how `Object` graphs are represented, and can ultimately be
+queried, consider how the hardware instance graph is elaborated. To build the
+`Object` graph, we first pick an entrypoint module to start elaboration. The
+elaboration process works according to the Verilog spec's definition of
+elaboration--instances of modules and `Object`s are instantiated in-memory,
+with connections to their inputs and outputs. Inputs are supplied, and outputs
+may be read. After elaboration completes, the `Object` graph is exposed in terms
+of the output ports, which may contain any `Property` types, including
+references to `Object`s.
+
+To illustrate how these pieces come together, consider the following examples:
+
+```scala mdoc:silent
+import chisel3.properties.Class
+import chisel3.experimental.hierarchy.{instantiable, public, Definition, Instance}
+
+// An abstract description of a CSR, represented as a Class.
+@instantiable
+class CSRDescription extends Class {
+  // An output Property indicating the CSR name.
+  val identifier = IO(Output(Property[String]()))
+  // An output Property describing the CSR.
+  val description = IO(Output(Property[String]()))
+  // An output Property indicating the CSR width.
+  val width = IO(Output(Property[Int]()))
+
+  // Input Properties to be passed to Objects representing instances of the Class.
+  @public val identifierIn = IO(Input(Property[String]()))
+  @public val descriptionIn = IO(Input(Property[String]()))
+  @public val widthIn = IO(Input(Property[Int]()))
+
+  // Simply connect the inputs to the outputs to expose the values.
+  identifier := identifierIn
+  description := descriptionIn
+  width := widthIn
+}
+```
+
+The `CSRDescription` is a `Class` that represents domain-specific information
+about a CSR. This uses `@instantiable` and `@public` so the `Class` can work
+with the `Definition` and `Instance` APIs.
+
+The readable fields we want to expose on `Object`s of the `CSRDescription` class
+are a string identifier, a string description, and an integer bitwidth, so these
+are output `Property` type ports on the `Class`.
+
+To capture concrete values at each `Object` instantiation, we have corresponding
+input `Property` type ports, which are connected directly to the outputs. This
+is how we would represent something like a Scala `case class` using `Class`.
+
+```scala mdoc:silent
+// A hardware module representing a CSR and its description.
+class CSRModule(
+  csrDescDef:     Definition[CSRDescription],
```

Comment:
Wondering why passing `Definition[CSRDescription]` as a parameter `csrDescDef:     Definition[CSRDescription]`?

Reply:
There are a couple annoying reasons. The short answer is that this is the way I'd write it today, so I'm trying to show the happy(est) path on the example.

Slightly longer answer... For one, we want to get the type of a reference to a CSRDescription in the Top module for the `definitions` IO, so it is useful to make the Definition there. Unfortunately, with the way D/I works today, if we also made new Definitions or used Instantiate within the CSRModule class, it will make duplicate definitions. I've spent some time trying to figure this out, but it in general it seems like Definitions are not shared across multiple calls to `Module`. So, the best way to get the FIRRTL we want, with one CSRDescription, is to only call Definition once, and pass the Definition to where it's needed. If you don't mind creating duplicate Definitions, it's not a problem, but I really wanted to show FIRRTL with one Definition of CSRDescription.

Reply:
I see, the reason I ask this is I wanna Module is able to serialized and de-serilaized, when add a `Definition` in `Module`, it's really hard to pickle it. I think we can still use the traditional way to instantiate a `Object` w/o passing `CSRDescription` as parameter, I'll tweak it.